### ISSUE TITLE
Make Che Theia webviews work in Che

### ISSUE THREAD

**Issue by mmorhun** [eclipse-che/che#15283]
There is a [PR](https://github.com/eclipse-theia/theia/pull/6465) in Eclipse Theia which reworks completely webview mechanism.

Some plugins are using webviews (like the Che-Theia welcome page, dependency analytics plugin and many others) and they will be broken (webview content will be empty) once the PR is merged.

According to the new model each webview will have a separate host (usually subdomain of main instance). We need to adapt Che Theia to this changes to make new webviews work in Che.

**Comment by skabashnyuk** [comment]
Could you please also check that this feature also works on single-host mode.

**Comment by mmorhun** [comment]
I will try. But last time I had issues with deploying single-host mode, so @skabashnyuk could you please provide precise steps how I can deploy Che multiuser with TLS (self signed cert is ok) single-host instance. Let's suppose that the first step is `minikube delete`. Thanks.

**Comment by tsmaeder** [comment]
@mmorhun is it correct to assume that extensions that work in VS Code will work with the new code in Theia? Do we know the places we need to fix in che-theia? How can we tell when our code is broken?

**Comment by mmorhun** [comment]
@tsmaeder yes VSCode extensions with webview should work in Eclipse Theia in case of using localhost (Otherwise https is required see [this comment](https://github.com/eclipse-theia/theia/pull/6465#issuecomment-557494879)).
As for now, when new image of Che Theia will be built by our CI, all webviews won't work (For example, welcome page is not displayed).
@AndrienkoAleksandr and I are investigating ways to resolve the problem for Che Theia.

**Comment by mmorhun** [comment]
For now we are going to set `THEIA_WEBVIEW_EXTERNAL_ENDPOINT` to the current Che Theia route to have it work as it worked before (no security, though). 
Then, to improve security, we'll try to use another route for all webviews (so Theia and all webviews will be separated, but no separation between webviews).

**Comment by benoitf** [comment]
We might stick to an upstream commit before rewrite of webview support if it takes more than one day to handle problem.

**Comment by tsmaeder** [comment]
Note that we need that reworked welcome support for Quarkus tooling: https://github.com/eclipse/che/issues/15113. So working around is ok for the moment, but we'll need a solution in the near future.

**Comment by mmorhun** [comment]
After merging of https://github.com/eclipse/che-theia/pull/563 webviews are working again (on the same domain, so no separation (security issue), but this is how it was before) but with new requirement to have Che Theia deployed using SSL (htttps protocol is required).

**Comment by mmorhun** [comment]
@skabashnyuk could you please provide step by step instruction how to deploy Che in single host mode with SSL? Otherwise I am blocked with single host problem.

**Comment by AndrienkoAleksandr** [comment]
> @skabashnyuk could you please provide step by step instruction how to deploy Che in single host mode with SSL? Otherwise I am blocked with single host problem.

I believe we should create docs for this stuff to share it with everyone...

**Comment by skabashnyuk** [comment]
https://docs.google.com/document/d/16OUQgXFuWDeH2Fu-je2CVD40JuFcBHbHsO0w5veDo0c/edit

**Comment by mmorhun** [comment]
To accomplish host separation, I've added another endpoint into Theia editor yaml (in my test environment) which points to the same port:
```yaml
...
spec:
  endpoints:
   -  name: "theia"
      public: true
      targetPort: 3100
      attributes:
        protocol: http
        type: ide
        secure: true
        cookiesAuthEnabled: true
        discoverable: false
   -  name: "webviews"
      public: true
      targetPort: 3100
      attributes:
        protocol: http
        type: webview
        secure: true
        cookiesAuthEnabled: true
        discoverable: false
...
```
but Che Master creates only one route for it. In other words, in workspace configuration I see two servers which has identical url, which makes impossible described above approach with separation of Che Theia and webviews hosts. So we need a way to say Che Mater to have two routes to the same target port in workspace pod.

**Comment by sparkoo** [comment]
cc: @skabashnyuk ^^

**Comment by mmorhun** [comment]
I faced new problem in separating webviews from Che-Theia. Beside https://github.com/eclipse/che/issues/15430 we have problem with workspace loader.

To explain the problem I must elaborate on some routes details first. Each route in workspace has attributes. Let's consider attributes related to security. They are `secure` and `cookiesAuthEnabled`. When `secure` attribute is set to false this means that the route is completely open and anyone can access it (for this case `cookiesAuthEnabled` doesn't make sense). If `secure` is true, Che puts JWT proxy behind route and the proxy allows only authenticated queries. A query considered as authenticated if the request has corresponding token. To automate the process of sending token, there is a mechanism which sets auth cookie if it isn't present and redirects to the target url. This mechanism could be turned on by `cookiesAuthEnabled` attribute of the route. See more in [docs](https://www.eclipse.org/che/docs/che-7/making-a-workspace-portable-using-a-devfile/#endpoints).

In case of webviews which are hosted on another from Che-Theia route we have problem with loader (which implements setting the auth cookie). If we have set `cookiesAuthEnabled` then the request from Theia to webview will fail because Theia will get loader code instead of expected response (this works for user in browser as script on loader page sets the cookie and redirects to the target url). If we turn off `cookiesAuthEnabled` flag, Theia will fail to get the resources as its queries miss the auth token (cookie). As workaround we may get the token from Che-Theia in some way and set to such requests (which is not good solution), but we cannot do the same for webview content and all requests from webview to their resources will fail (as no one is able to set the token inside plugin code).

@skabashnyuk have I missed something? Do you have an idea what we can do here?

**Comment by mmorhun** [comment]
I faced new problem in separating webviews from Che-Theia. Beside https://github.com/eclipse/che/issues/15430 we have problem with workspace loader.

To explain the problem I must elaborate on some routes details first. Each route in workspace has attributes. Let's consider attributes related to security. They are `secure` and `cookiesAuthEnabled`. When `secure` attribute is set to false this means that the route is completely open and anyone can access it (for this case `cookiesAuthEnabled` doesn't make sense). If `secure` is true, Che puts JWT proxy behind the route and JWT proxy allows only authenticated queries. A query considered as authenticated if the request has corresponding token (cookie, header or query parameter with the token). To automate the process of sending token, there is a mechanism that redirects to special page (loader) which sets auth cookie if user are logged in with Keycloak. Then the loader redirects to the target url (and now the auth cookie is set). This mechanism could be turned on by `cookiesAuthEnabled` attribute of the route. See more in [docs](https://www.eclipse.org/che/docs/che-7/making-a-workspace-portable-using-a-devfile/#endpoints).

In case of webviews which are hosted on another from Che-Theia route we have problem with loader (which implements setting the auth cookie). If we have set `cookiesAuthEnabled` then the request from Theia to webview will fail because Theia (webviews) will get loader code instead of expected response (this works for user in browser as script on loader page sets the cookie and redirects to the target url, which is not possible for just a query). If we turn off `cookiesAuthEnabled` flag, Theia will fail to get the resources as its queries miss the auth token (cookie). As an attempt to workaround that we may get the token from Che-Theia in some way and set to such requests (which is not good solution), but we cannot do the same for webview content and all requests from webview to their resources will fail (as no one is able to set the token inside plugin code).

@skabashnyuk have I missed something? Do you have an idea what we can do here?

**Comment by mmorhun** [comment]
I also tried workaround with hidden iframe which does dummy request to webviews route what should set the cookie for Theia (webviews frontend host). But it failed with CORS blocking:
![cors](https://user-images.githubusercontent.com/15607393/70782786-7abe2800-1d90-11ea-83a8-16f10aa858d5.png)

If set `sequre` flag for the routes to false then everything will work but without any security, of course.

**Comment by mmorhun** [comment]
Also forgot to mention, if open webviews route manually in browser, it will set the auth cookie and after Che-Theia page refresh webviews are working.

**Comment by azatsarynnyy** [comment]
I'm closing it as Theia WebViews are working in Che Theia, on the same origin.
Opened a follow-up issue to track the work for isolating the WebViews from Theia page https://github.com/eclipse/che/issues/15616
